### ISSUE TITLE
[CUDA] torch.sum over smaller inner dimension is 4x slower than it should be

### ISSUE THREAD

**Issue by jerrybai1995** [pytorch/pytorch#18807]
(Edited by @colesbury to fix benchmarking)

Simpler repro:

-----

```python
res = torch.rand(7840000, 4, device='cuda')
%timeit res[:, 0] + res[:, 1] + res[:, 2] + res[:, 3]; torch.cuda.synchronize()
%timeit res.sum(dim=-1); torch.cuda.synchronize()
```

------

Hi,

I found a strange behavior:

```python
temp1 = torch.rand(4,1400,1400,1,4, device='cuda')
temp2 = torch.rand(4,1400,1400, device='cuda')
def h():
    res = torch.einsum('bijld, dij -> bijld', temp1, temp2)
    return res[:,:,:,:,0] + res[:,:,:,:,1] + res[:,:,:,:,2] + res[:,:,:,:,3]
%timeit h(); torch.cuda.synchronize()

def p():
    res = torch.einsum('bijld, dij -> bijld', temp1, temp2)
    return torch.sum(res, dim=4)
%timeit p(); torch.cuda.synchronize()
```
Output:
```
1.66 ms ± 459 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
4.54 ms ± 20.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Why is `h()` significantly faster than `p()`? I'm very confused.

**Comment by colesbury** [comment]
Is this on CPU (did you change the default tensor type to GPU)?

Please follow the steps from the issue tracker template. (What PyTorch version are you using?, etc. [collect_env.py](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py))

**Comment by jerrybai1995** [comment]
Hi @colesbury ,

This is on GPU, the default tensor type is `torch.cuda.FloatTensor`. 

```
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti

Nvidia driver version: 418.40.04
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip3] numpy==1.15.3
[conda] blas                      1.0                         mkl
[conda] cuda100                   1.0                           0    pytorch
[conda] cuda80                    1.0                           0    soumith
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] magma-cuda100             2.5.0                         1    pytorch
[conda] mkl                       2019.1                      144
[conda] mkl-service               1.1.2            py36he904b0f_5
[conda] mkl_fft                   1.0.10           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.0.0           py3.6_cuda10.0.130_cudnn7.4.1_1  [cuda100]  pytorch
[conda] torchtext                 0.4.0                    pypi_0    pypi
[conda] torchvision               0.2.1                      py_2    pytorch
```

**Comment by colesbury** [comment]
@jerrybai1995 thanks for the report. You need to use `torch.cuda.synchronize()` when benchmarking CUDA code. Removing the einsum and simplifying the code leads to:

```python
res = torch.rand(7840000, 4, device='cuda')
%timeit res[:, 0] + res[:, 1] + res[:, 2] + res[:, 3]; torch.cuda.synchronize()
%timeit res.sum(dim=-1); torch.cuda.synchronize()
```

On my GPU (P100) with PyTorch master this gives 1.01 ms and 1.19 ms. Your results might vary slightly.

It does look like sum() is slower than it should be. On my GPU with ~500GB/sec of memory bandwidth (measured by timing clone()) I'd expect good performance to be:

```
(7840000 * 4 * 4 bytes + 7840000 * 4 bytes) / (500 GB/sec) = 313 µs
[input size + output size] / [memory bandwidth]
```

**Comment by jerrybai1995** [comment]
Hi @colesbury, 

I just benchmarked your simplified code too, on 2080ti:
```python
res = torch.rand(7840000, 4, device='cuda')
%timeit res[:, 0] + res[:, 1] + res[:, 2] + res[:, 3]; torch.cuda.synchronize()
%timeit res.sum(dim=-1); torch.cuda.synchronize()
```
The output is:
```
999 µs ± 489 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
3.8 ms ± 12.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
which is still about 4x slower. I have made sure that no other process is using the GPU. Do you think it's the problem of the type of GPU (2080ti vs P100)?

**Comment by colesbury** [comment]
@jerrybai1995 I don't have easy access to a 2080ti. Can you also run the following:

```python
res = torch.rand(7840000, 4, device='cuda')
%timeit res.clone(); torch.cuda.synchronize()
%timeit res.sum(); torch.cuda.synchronize()
```

I'd like to estimate the realizable bandwidth and I'm curious if the sum() over all dimensions is performant (it appears to perform well on my GPU)

**Comment by jerrybai1995** [comment]
Here is the output:
```
481 µs ± 47.9 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
237 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by ezyang** [comment]
@colesbury please feel free to unassign yourself or deprioritize as necessary.

**Comment by colesbury** [comment]
Ohhh, I don't think it's P100 vs 2080 TI -- I think it's PyTorch 1.0.0 vs. master. I installed PyTorch 1.0.0 on my machine and the sum(-1) takes ~9 ms. (Same with PyTorch 1.0.1). If you'd like you can try the nightly build from https://pytorch.org/.

I think @jjsjann123 sped it up since PyTorch 1.0.

Still, the reduction is ~4x slower than it should be. In this case (sum over inner dimension), we parallelize over the 32 threads in a warp.

https://github.com/pytorch/pytorch/blob/0916b5419a5703f7f81822b66f31642c0613ccad/aten/src/ATen/native/cuda/Reduce.cuh#L661

In this case, the warp should probably work on multiple output items.

**Comment by jerrybai1995** [comment]
So do you mean it's probably just better to not use reduction on small inner dimensions?

I thought it might be because the memory bandwidth issue too, but it's weird that both `.clone()` and `.sum()` are much faster than `.sum(-1)`. Will the speedup update be pushed to the next version of PyTorch? (I'll certainly try the nightly build in the meantime)

**Comment by jjsjann123** [comment]
Sum along a small inner dimension has got a decent speedup from launch config PR that currently in the nightly.

I saw decent speedup (~6x)from 1.0 to nightly running the following `torch.rand(7840000, 4, device='cuda').sum(dim=-1)`

**Comment by mruberry** [comment]
@jerrybai1995 Are you still seeing slow performance even with a more recent version of PyTorch (one within the past 20 days)?
